### PR TITLE
SuSE Linux 15 support

### PR DESCRIPTION
**Pull request by jjohnson42**
Diskful install of SLE 15 support

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/sles.pm [hunk 1]
```diff
@@ -1740,16 +1742,16 @@ sub copycd
...
-    my $dsc = <$dinfo>;
-    if ($dsc =~ /x86_64/) {
-        $darch = "x86_64";
-    }
-    if ($dsc =~ /Installer/ and $dsc =~ /SLE-/) {
+        my $dsc = <$dinfo>;
+        if ($dsc =~ /x86_64/) {
+            $darch = "x86_64";
+        }
+        if ($dsc =~ /Installer/ and $dsc =~ /SLE-15/) {
```
Comment (bybai) [inline xCAT-server/lib/xcat/plugins/sles.pm]: From line 1749 to line 1754, you change "sle" into "sle15", I think this code logic should work for other sles version, if change "sle" into "sle15", it becomes only support "sle15", does not cause problem for old sle version?

Comment (jjohnson42) [reply xCAT-server/lib/xcat/plugins/sles.pm]: The problem with the sles12 was caused by discnumber scope problem...  The most recent commit fixes the discnumber scope mistake.

In SuSE 15, all enterprise editions share common media, so user is no longer copying 'sles' specifically in.

In fact, the compute template here actually installs SLE HPC, not SLES.

File: xCAT-server/lib/xcat/plugins/anaconda.pm [hunk 2]
```diff
@@ -335,9 +335,15 @@ sub mknetboot
...
-            next; 
+            my @ents = xCAT::TableUtils->get_site_attribute("disablenodesetwarning");
+            my $site_ent = $ents[0];
+            if (!defined($site_ent) || ($site_ent =~ /no/i) || ($site_ent =~ /0/))
+            {
+                if (!defined($::DISABLENODESETWARNING)) {    # set by AAsn.pm
+                    xCAT::MsgUtils->report_node_error($callback, $node, "OS image name must be specified in nodetype.provmethod");
+                    next;
+		}
+            }
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/anaconda.pm]: @jjohnson42 the `next` should be here,  xcat is not supporting the legacy node. With the `next`, the following codes will not be executed.

File: xCAT-server/lib/xcat/plugins/anaconda.pm [hunk 3]
```diff
@@ -1064,9 +1070,16 @@ sub mkinstall
...
+
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/anaconda.pm]: `next` should be moved out to short stop the below codes.

File: xCAT-server/lib/xcat/plugins/sles.pm [hunk 4]
```diff
@@ -203,8 +203,15 @@ sub mknetboot
...
-            next;
+            my @ents = xCAT::TableUtils->get_site_attribute("disablenodesetwarning");
+            my $site_ent = $ents[0];
+            if (!defined($site_ent) || ($site_ent =~ /no/i) || ($site_ent =~ /0/))
+            {
+                if (!defined($::DISABLENODESETWARNING)) {    # set by AAsn.pm
+                    xCAT::MsgUtils->report_node_error($callback, $node, "OS image name must be specified in nodetype.provmethod");
+                    next;
+                }
+            }
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/sles.pm]: next should be moved out to short stop the below deprecated codes.

File: xCAT-server/lib/xcat/plugins/sles.pm [hunk 5]
```diff
@@ -849,15 +856,22 @@ sub mkinstall
...
-
+            my @ents = xCAT::TableUtils->get_site_attribute("disablenodesetwarning");
+            my $site_ent = $ents[0];
+            if (!defined($site_ent) || ($site_ent =~ /no/i) || ($site_ent =~ /0/))
+            {
+                if (!defined($::DISABLENODESETWARNING)) {    # set by AAsn.pm
+                    xCAT::MsgUtils->report_node_error($callback, $node, "OS image name must be specified in nodetype.provmethod");
+                    next;
+                }
+            }
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/sles.pm]: `next` should be moved out to short stop the below deprecated codes.

File: xCAT-server/share/xcat/install/sles/compute.sle15.tmpl [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      <!-- XCAT-PARTITION-END -->
+    </partitioning>
+    <add-on>
+        <add_on_products config:type="list">
+            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sles15/x86_64/2</media_url><product>sle-module-basesystem</product><product_dir>/Module-Basesystem</product_dir></listentry>
+            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sles15/x86_64/2</media_url><product>sle-module-hpc</product><product_dir>/Module-HPC</product_dir></listentry>
+            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sles15/x86_64/2</media_url><product>sle-module-server-applications</product><product_dir>/Module-Server-Applications</product_dir></listentry>
+            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sles15/x86_64/2</media_url><product>sle-module-server-applications</product><product_dir>/Module-Server-Applications</product_dir></listentry>
+            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sles15/x86_64/2</media_url><product>SLE_HPC</product><product_dir>/Product-HPC</product_dir></listentry>
+        </add_on_products>
```
Comment (robin2008) [inline xCAT-server/share/xcat/install/sles/compute.sle15.tmpl]: Is the `add-on` section here is mandatory?  And it is not good to hard code `x86_64` in the general template.

File: xCAT-server/lib/xcat/plugins/anaconda.pm (hunk 2 above)
Comment (robin2008) [reply xCAT-server/lib/xcat/plugins/anaconda.pm]: Okay, I see your commit comments. It seems you really want bring the legacy mode back to work with `disablenodesetwarning`

**Comment by robin2008** [comment]
@bybai  Let's merge it first after run a UT for sles12.3 ppc64le copycds and provision.

**Comment by bybai** [comment]
@robin2008, our autotest sles12sp3 diskful installation test case is passed when it is using this "sles.pm", so we can let auto test to test it first.

**Comment by bybai** [comment]
I am OK to merge it.

**Pull request merged**